src/useSubscription.ts

```diff
@@ -0,0 +1,86 @@
+import {
+  ApolloClient,
+  ApolloError,
+  OperationVariables,
+  SubscriptionOptions,
+} from 'apollo-client';
+import { DocumentNode } from 'graphql';
+import { useEffect, useRef, useState } from 'react';
+import { useApolloClient } from './ApolloContext';
+import { Omit, objToKey } from './utils';
+
+export type OnSubscriptionData<TData> = (
+  options: OnSubscriptionDataOptions<TData>
+) => any;
+
+export interface OnSubscriptionDataOptions<TData> {
+  client: ApolloClient<any>;
+  subscriptionData: SubscriptionHookResult<TData>;
+}
+
+export interface SubscriptionHookOptions<TData, TVariables>
+  extends Omit<SubscriptionOptions<TVariables>, 'query'> {
+  onSubscriptionData?: OnSubscriptionData<TData>;
+}
+
+export interface SubscriptionHookResult<TData> {
+  data?: TData;
+  error?: ApolloError;
+  loading: boolean;
+}
+
+export function useSubscription<TData = any, TVariables = OperationVariables>(
+  query: DocumentNode,
+  {
+    variables,
+    onSubscriptionData,
+    ...options
+  }: SubscriptionHookOptions<TData, TVariables> = {}
+): SubscriptionHookResult<TData> {
+  const client = useApolloClient();
+  const onSubscriptionDataRef = useRef<null | OnSubscriptionData<TData>>(null);
+  const [result, setResult] = useState<SubscriptionHookResult<TData>>({
+    loading: true,
+  });
+
+  if (onSubscriptionData) {
+    onSubscriptionDataRef.current = onSubscriptionData;
+  }
+
+  useEffect(
+    () => {
+      const subscription = client
+        .subscribe({
+          ...options,
+          query,
+          variables,
+        })
+        .subscribe({
```

Comment:
`error` before `next` looks a little unusual for me. I guess it's that way because of a tslint rule. So I'd suggest using a different form of `subscribe` here:

```
       .subscribe(
          nextResult => {
            // ...
          },
          error => {
            // ...
          }
        );
```

It's also the same form used in `useQuery` and it should allow us to save a byte or two in the final bundle 😉.